src/util/locking/RedisReadWriteLocker.ts

```diff
@@ -0,0 +1,158 @@
+   */
+  private createRedisClient(redisClientString: string): Redis {
+    if (redisClientString.length > 0) {
+      // Check if port number or ip with port number
+      // Definitely not perfect, but configuring this is only for experienced users
+      const match = new RegExp(/^(?:([^:]+):)?(\d{4,5})$/u, 'u').exec(redisClientString);
+      if (!match || !match[2]) {
+        // At least a port number should be provided
+        throw new Error(`Invalid data provided to create a Redis client: ${redisClientString}\n
+            Please provide a port number like '6379' or a host address and a port number like '127.0.0.1:6379'`);
+      }
+      const port = Number(match[2]);
+      const host = match[1];
+      return new Redis(port, host);
+    }
+    throw new Error(`Empty redisClientString provided!\n
+            Please provide a port number like '6379' or a host address and a port number like '127.0.0.1:6379'`);
+  }
+
+  public async withReadLock<T>(identifier: ResourceIdentifier, whileLocked: () => (Promise<T> | T)): Promise<T> {
+    const key = this.getKey(identifier);
+    const readLock = await this.tryRedisFn((): Promise<boolean> => fromResp2ToBool(this.redis.acquireReadLock(key)));
+    if (readLock) {
+      try {
+        return await whileLocked();
+      } finally {
+        await this.tryRedisFn((): Promise<boolean> => fromResp2ToBool(this.redis.releaseReadLock(key)));
+      }
+    } else {
+      throw new Error('Error trying to get readlock.');
+    }
+  }
+
+  public async withWriteLock<T>(identifier: ResourceIdentifier, whileLocked: () => (Promise<T> | T)): Promise<T> {
+    const key = this.getKey(identifier);
+    const writeLock = await this.tryRedisFn((): Promise<boolean> => fromResp2ToBool(this.redis.acquireWriteLock(key)));
+    if (writeLock) {
+      try {
+        return await whileLocked();
+      } finally {
+        await this.tryRedisFn((): Promise<boolean> => fromResp2ToBool(this.redis.releaseWriteLock(key)));
+      }
+    } else {
+      throw new Error('Error trying to get writelock.');
+    }
+  }
+
+  /**
+   * Try a redis function according to the set {@link AttemptSettings}
+   *
+   * @param fn - The function to try
+   *
+   * @returns A promise wrapped boolean
+   *
+   * @see To convert from redis operation to Promise<boolean> use {@link fromResp2ToBool} to wrap the function
+   */
+  private async tryRedisFn(fn: () => Promise<boolean>): Promise<boolean> {
+    const settings = this.attemptSettings;
+    const maxTries = settings.retryCount === -1 ? Number.POSITIVE_INFINITY : settings.retryCount + 1;
+    function calcTime(): number {
+      return Math.max(0, settings.retryDelay + Math.floor(Math.random() * settings.retryJitter));
+    }
+
+    let tries = 1;
+    let acquired = await fn();
+    // Keep going until either you get a lock/release or maxTries has been reached.
+    while (!acquired && (tries <= maxTries)) {
```

Comment:
`tryRedisRn(fn)` will execute said fn. It will await and if the fn succeeds (in acquireing or release read/write) it will return true, if it does not succees (because a lock is currently held by another process) it will return false. Any errors that are thrown are not part of the regular lock logical flow. So variant of your first case, not the second case.

Reply:
Would add some more comments then to explain that this function exists because Redis doesn't have some sort of waiting system for locks.

Reply:
I've simplified that function a little bit to now return Promise<void> en thus return if the operation succeeds. It will reject with an error otherwise. It is now explained in the jsdoc above it too.